Source/WTF/wtf/FileHandle.cpp

```diff
@@ -0,0 +1,158 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FileHandle.h"
+
+#include <wtf/FileSystem.h>
+
+namespace WTF::FileSystemImpl {
+
+FileHandle::FileHandle() = default;
+
+FileHandle::FileHandle(FileHandle&& other)
+    : m_handle(std::exchange(other.m_handle, std::nullopt))
+#if USE(FILE_LOCK)
+    , m_isLocked(std::exchange(other.m_isLocked, false))
+#endif
+{ }
+
+FileHandle::~FileHandle()
+{
+    closeIfNecessary();
+}
+
+FileHandle& FileHandle::operator=(FileHandle&& other)
+{
+    closeIfNecessary();
+    m_handle = std::exchange(other.m_handle, std::nullopt);
+#if USE(FILE_LOCK)
+    m_isLocked = std::exchange(other.m_isLocked, false);
+#endif
+    return *this;
+}
+
+std::optional<Vector<uint8_t>> FileHandle::readAll()
+{
+    if (!isValid())
+        return { };
+
+    auto size = this->size().value_or(0);
+    if (!size)
+        return std::nullopt;
+
+    size_t bytesToRead;
+    if (!WTF::convertSafely(size, bytesToRead))
+        return std::nullopt;
+
+    Vector<uint8_t> buffer(bytesToRead);
+    size_t totalBytesRead = 0;
+    int bytesRead;
+
+    while ((bytesRead = read(buffer.mutableSpan().subspan(totalBytesRead))) > 0)
+        totalBytesRead += bytesRead;
+
+    if (totalBytesRead != bytesToRead)
+        return std::nullopt;
+
+    return buffer;
+}
+
+int FileHandle::read(std::span<uint8_t> data)
+{
+    if (!isValid())
+        return -1;
+    return readFromFile(platformHandle(), data);
+}
+
+int FileHandle::write(std::span<const uint8_t> data)
+{
+    if (!isValid())
+        return -1;
+    return writeToFile(platformHandle(), data);
+}
+
+void FileHandle::lock(OptionSet<FileLockMode> lockMode)
+{
+#if USE(FILE_LOCK)
+    RELEASE_ASSERT(!m_isLocked);
+    m_isLocked = FileSystemImpl::lockFile(platformHandle(), lockMode);
+    ASSERT(m_isLocked);
+#else
+    UNUSED_PARAM(lockMode);
+#endif
+}
+
+bool FileHandle::truncate(long long offset)
+{
+    if (!isValid())
+        return false;
+    return truncateFile(platformHandle(), offset);
+}
+
+void FileHandle::closeIfNecessary()
+{
+    if (!isValid())
+        return;
+
+    auto handle = std::exchange(m_handle, std::nullopt).unsafeValue();
```

Comment:
Since we're paying for filesystem operations anyway, we might as well use `value()` and drop the unsafety. Also, the `isValid()` check is trivially optimizable, so the compiler will just de-duplicate the code for us.